filestorage.go

```diff
@@ -149,125 +152,124 @@ func dataDir() string {
 	return filepath.Join(baseDir, "certmagic")
 }
 
-// TryLock attempts to get a lock for name, otherwise it returns
-// a Waiter value to wait until the other process is finished.
-func (fs FileStorage) TryLock(key string) (Waiter, error) {
-	fileStorageNameLocksMu.Lock()
-	defer fileStorageNameLocksMu.Unlock()
+// Lock obtains a lock named by the given key. It blocks
+// until the lock can be obtained or an error is returned.
+func (fs *FileStorage) Lock(key string) error {
+	// can't defer the unlock because we might have
+	// to Wait() for a while before returning, so we're
+	// careful to unlock at all the right places
+	fs.fileStorageNameLocksMu.Lock()
+
+	if fs.fileStorageNameLocks == nil {
+		fs.fileStorageNameLocks = make(map[string]*fileStorageWaiter)
+	}
 
 	// see if lock already exists within this process - allows
 	// for faster unlocking since we don't have to poll the disk
-	fw, ok := fileStorageNameLocks[key]
+	fw, ok := fs.fileStorageNameLocks[key]
 	if ok {
 		// lock already created within process, let caller wait on it
-		return fw, nil
+		fs.fileStorageNameLocksMu.Unlock()
+		fw.Wait()
+		return nil
 	}
 
 	// attempt to persist lock to disk by creating lock file
 
-	// parent dir must exist
 	lockDir := fs.lockDir()
-	if err := os.MkdirAll(lockDir, 0700); err != nil {
-		return nil, err
-	}
-
+	// since there isn't already a waiter for the lock, make one
 	fw = &fileStorageWaiter{
 		key:      key,
 		filename: filepath.Join(lockDir, StorageKeys.safe(key)+".lock"),
 		wg:       new(sync.WaitGroup),
 	}
+	fw.wg.Add(1)
+	fs.fileStorageNameLocks[key] = fw
+	fs.fileStorageNameLocksMu.Unlock()
 
-	var checkedStaleLock bool // sentinel value to avoid infinite goto-ing
+	for {
+		// parent dir must exist
+		if err := os.MkdirAll(lockDir, 0700); err != nil {
+			return err
+		}
 
-createLock:
-	// create the file in a special mode such that an
-	// error is returned if it already exists
-	lf, err := os.OpenFile(fw.filename, os.O_CREATE|os.O_EXCL, 0644)
-	if err != nil {
-		if os.IsExist(err) {
+		// create the file in a special mode such that an
+		// error is returned if it already exists
+		lf, err := os.OpenFile(fw.filename, os.O_CREATE|os.O_EXCL, 0644)
+		switch {
+		case err == nil:
+			// cool, we got the lock right away
+			lf.Close()
+			return nil
+		case os.IsExist(err):
 			// another process has the lock
-
-			// check to see if the lock is stale, if we haven't already
-			if !checkedStaleLock {
-				checkedStaleLock = true
-				if fs.lockFileStale(fw.filename) {
-					log.Printf("[INFO][%s] Lock for '%s' is stale; removing then retrying: %s",
-						fs, key, fw.filename)
-					os.Remove(fw.filename)
-					goto createLock
-				}
+			info, err := os.Stat(fw.filename)
```

Comment:
I noticed that this now leaves the `lockFileStale()` method unused -- was this intentional? I am ready to push a small commit to restore this if that is desirable.

Reply:
I didn't use `lockFileStale()` because it returns `true` if stat fails (which isn't the same as the file existing being stale).

Reply:
Well, that function returning `true` has the same net result as this change: the loop will simply continue.

I'll push a commit in a moment that adjusts based on this feedback, plus a little of my own yak shaving :smile:

Reply:
p2 cannot create its own file if p1 already created it: the file creation is atomic (assuming the file system is correctly implemented) because of the O_EXCL flag. Thus p2 will not create its own file and instead wait for the lock to be released.

Reply:
I don't think that will work either unless the Chtimes call is atomic with the creation of the file...

Reply:
I know this is less than ideal, but at least consider this:

Since certificate operations are very infrequent (currently once about every 60 days, only lasting a few seconds), and the consequences of a highly improbable (admittedly not impossible) race are essentially nil (in the event of a concurrent issuance or renewal, Let's Encrypt will still allow both -- it's just if it happens often for a higher number of subdomains that they throttle), maybe the locking doesn't have to be perfect, honestly? And even if the rate limit _is_ hit (which I already think is very unlikely), it would be even _less_ likely the next time, and would still result in no downtime since renewals can be tried for the next 3-4 weeks.

Just something to consider.

But of course I'm down to make this as perfect as possible (within practicality).

Reply:
I *think* `Unlock` suffers from the same `os.Remove` race we've seen elsewhere in the face of a stale lock - as unlikely as it is:

* p2 sees the lock is stale
* p2 removes it
* p2 creates its own lock
* p1 completes
* p1 removes the file
* p3 is free to proceed

I think a better strategy for the plain `os.Remove` is:

* create another file e.g. `key.unlock` first
* if it succeeds, we take ownership of unlocking:
*  * remove key.lock
*  * remove key.unlock
* otherwise, another process is doing the unlocking

There is a very small risk that the app crashes between `remove key.lock` and `key.unlock` so `Lock` would need to remove `key.unlock` when it successfully creates `key.lock`.

And now you know why I'm not a fan of `lock files`.

Reply:
I agree the locking doesn't need to be perfect (given the use-case), so don't let me be a blocker :). Considering I appear to be the first one to bring up the locking issues, it probably means it's not a big problem.

Reply:
I think I understand. I'm willing to accept the infinitesimally small risk_ that the program will crash in that critical section, because even then the consequences won't even be tangible, I think.

Do you want to try implementing those changes? If so, I will hold off! (In fact I have to go do my thesis work for the day, so I won't get around to it for a bit anyway.)

Reply:
I'll add it to my list for tomorrow.